### PR TITLE
Disable persistence of completed job metrics by default

### PR DESCRIPTION
**Pull request by cangencer**
New option `JobConfig.setStoreMetricsAfterJobCompletion` 
can be used to control whether metrics are saved upon
job completion.

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java [hunk 1]
```diff
@@ -400,39 +391,34 @@ public void reset() {
...
-                    }
+                JobExecutionRecord record = jobRepository.getJobExecutionRecord(jobId);
+                if (record != null) {
+                    // job is suspended, or just submitted now
+                    cf.complete(JobMetrics.empty());
+                    return;
+                }
 
-                    cf.completeExceptionally(new JobNotFoundException(jobId));
+                JobResult jobResult = jobRepository.getJobResult(jobId);
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java]: Also we can do this the other way: first check JobMetrics, if not found, then check JobResult before throwing. It's unlikely the user will check for jobMetrics if he disabled it.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java [hunk 2]
```diff
@@ -636,9 +622,11 @@ boolean isQuorumPresent(int quorumSize) {
     CompletableFuture<Void> completeJob(MasterContext masterContext, long completionTime, Throwable error) {
         return submitToCoordinatorThread(() -> {
             // the order of operations is important.
-
             long jobId = masterContext.jobId();
-            JobMetrics jobMetrics = masterContext.jobContext().jobMetrics();
+            JobMetrics jobMetrics =
+                masterContext.jobConfig().isStoreMetricsAfterJobCompletion()
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java]: indetation...

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 3]
```diff
@@ -279,7 +280,11 @@ long newExecutionId() {
      * @throws IllegalStateException if the JobResult is already present
      */
     void completeJob(
-        long jobId, JobMetrics terminalMetrics, String coordinator, long completionTime, Throwable error
+        long jobId,
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: indentation

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/metrics/JobMetrics_BatchTest.java [hunk 4]
```diff
@@ -51,29 +55,39 @@
...
         // When
-        job.join();
+        Job job = execute(p, JOB_CONFIG_WITH_METRICS);
+
+        // Then
+        assertMetrics(job.getMetrics());
+    }
+
+    @Test
+    public void when_jobCompleted_without_savingMetricsOnCompletionEnabled_then_emptyMetrics() {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/metrics/JobMetrics_BatchTest.java]: `when_storeMetricsAfterJobCompletionDisabled_then_metricsEmpty`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/JetProperties.java [hunk 5]
```diff
@@ -29,13 +29,17 @@
...
+ * @since 3.2
  */
 public final class JetProperties {
 
     /**
      * Jet will periodically check for new jobs to start and perform cleanup of
      * unused resources. This property configures how often this check and
      * cleanup will be done.
+     *
+     * @since 3.2
```
[the remark this answers is not in the document]
Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/JetProperties.java]: Yes, but it's the first time it's part of public API. So we may consider it didn't exist before.

**Pull request merged**